src/legacy/reduce.rs

```diff
@@ -0,0 +1,486 @@
+
+use thiserror::Error;
+
+use crate::legacy::lz77::lz77_output_backref;
+
+use super::bitstream::{lsb, BitStream};
+
+#[derive(Error, Debug)]
+enum ReduceError {
+    #[error("Invalid follower set")]
+    InvalidFollowerSet,
+    #[error("Error reading next byte")]
+    ErrorReadingNextByte,
+}
+
+/// Number of bits used to represent indices in a follower set of size n.
+fn follower_idx_bw(n: u8) -> u8 {
+    assert!(n <= 32);
+
+    if n > 16 {
+        return 5;
+    }
+    if n > 8 {
+        return 4;
+    }
+    if n > 4 {
+        return 3;
+    }
+    if n > 2 {
+        return 2;
+    }
+    if n > 0 {
+        return 1;
+    }
+    return 0;
+}
+
+#[derive(Default, Clone, Copy)]
+struct FollowerSet {
+    size: u8,
+    idx_bw: u8,
+    followers: [u8; 32],
+}
+
+/// Read the follower sets from is into fsets. Returns true on success.
+fn read_follower_sets(is: &mut BitStream, fsets: &mut [FollowerSet]) -> bool {
+    for i in (0..=255 as usize).rev() {
+        let n = lsb(is.bits(), 6) as u8;
+        if n > 32 {
+            return false;
+        }
+        if !is.advance(6) {
+            return false;
+        }
+        fsets[i].size = n;
+        fsets[i].idx_bw = follower_idx_bw(n);
+
+        for j in 0..fsets[i].size as usize {
+            fsets[i].followers[j] = is.bits() as u8;
+            if !is.advance(8) {
+                return false;
+            }
+        }
+    }
+
+    return true;
+}
+
+/// Read the next byte from is, decoded based on prev_byte and the follower sets.
+/// The byte is returned in *out_byte. The function returns true on success,
+/// and false on bad data or end of input.
+fn read_next_byte(
+    is: &mut BitStream,
+    prev_byte: u8,
+    fsets: &mut [FollowerSet],
+    out_byte: &mut u8,
+) -> bool {
+    let bits = is.bits();
+
+    if fsets[prev_byte as usize].size == 0 {
+        // No followers; read a literal byte.
+        *out_byte = bits as u8;
+        return is.advance(8);
+    }
+
+    if lsb(bits, 1) == 1 {
+        // Don't use the follower set; read a literal byte.
+        *out_byte = (bits >> 1) as u8;
+        return is.advance(1 + 8);
+    }
+
+    // The bits represent the index of a follower byte.
+    let idx_bw = fsets[prev_byte as usize].idx_bw;
+    let follower_idx = lsb(bits >> 1, idx_bw) as usize;
+    if follower_idx >= fsets[prev_byte as usize].size as usize {
+        return false;
+    }
+    *out_byte = fsets[prev_byte as usize].followers[follower_idx];
+    return is.advance(1 + idx_bw);
+}
+
+fn max_len(comp_factor: u8) -> usize {
+    let v_len_bits = (8 - comp_factor) as usize;
+
+    assert!(comp_factor >= 1 && comp_factor <= 4);
+
+    // Bits in V + extra len byte + implicit 3.
+    ((1 << v_len_bits) - 1) + 255 + 3
+}
+
+fn max_dist(comp_factor: u8) -> usize {
+    let v_dist_bits = comp_factor as usize;
+
+    assert!(comp_factor >= 1 && comp_factor <= 4);
+
+    // Bits in V * 256 + W byte + implicit 1. */
+    ((1 << v_dist_bits) - 1) * 256 + 255 + 1
```

Comment:
This is equivalent to 
```suggestion
    1 << (v_dist_bits + 8)
```

Reply:
+1